_posts/2015-08-11-postgres-remove-duplicates.md

```diff
@@ -0,0 +1,168 @@
+in this set.
+
+We could achieve the same result with `DISTINCT ON`, let's see an example.
+
+### <a name="distinct_on"></a>DISTINCT ON
+
+{% highlight sql %}
+DELETE
+FROM orders
+WHERE id NOT IN (
+  SELECT DISTINCT ON (item_id, code) id
+  FROM orders
+);
+{% endhighlight %}
+
+Here we fetch the first row from each group with the same `item_id` and
+`name` and delete the rest. It's important to mention that we are relying in the
+order by `id`. If this is not the case we should explicitly order our result as
+mentioned [in the
+documentation](http://www.postgresql.org/docs/9.0/static/sql-select.html#SQL-DISTINCT).
+
+### <a name="exists"></a>EXISTS
+
+{% highlight sql %}
+DELETE
+FROM orders o1
+WHERE EXISTS (
+  SELECT 1
+  FROM orders o2
+  WHERE o1.item_id = o2.item_id AND
+        o1.code = o2.code AND
+        o1.id > o2.id
+);
+{% endhighlight %}
+
+This reads as: "Delete every row from o1 that has the same item_id and code and
+an id bigger than at least one id from o2". This will ensure that only the
+duplicates with the lowest id remains.
+
+### <a name="using"></a>USING
+
+{% highlight sql %}
+DELETE
+FROM orders o1 USING orders o2
+WHERE o1.item_id = o2.item_id AND
+      o1.code = o2.code AND
+      o1.id > o2.id;
+{% endhighlight %}
+
+Which executes the same code in our `GROUP BY` example. We also use the same
+technique explained above to keep the lowest id from our duplicated records.
+We can see another `USING` example
+[here](http://www.postgresql.org/docs/9.4/static/sql-delete.html#AEN78458).
+
+### <a name="partition"></a>PARTITION (WINDOW FUNCTION)
+
+{% highlight sql %}
+DELETE
+FROM orders
+WHERE id IN (
+  SELECT id
+  FROM (
+   SELECT row_number() OVER (PARTITION BY item_id, code ORDER BY id),
+          id
+   FROM orders
+  ) AS sorted
+  WHERE sorted.row_number > 1
+);
+{% endhighlight %}
+
+Here we partition our records using the specified criteria to find duplicates
+and sort this partition by `id`, after this we create a column with the row
+number for each row in a partition. The last step is to select only the rows
+with row numbers greater than 1 to delete.
+
+### <a name="temporary"></a>Temporary tables
+
+{% highlight sql %}
+CREATE TABLE temporary_orders AS
+  SELECT DISTINCT ON (item_id, code) * FROM orders;
+
+DROP TABLE orders;
+ALTER TABLE temporary_orders RENAME TO orders;
+{% endhighlight %}
+
+This solution is fast, but you have to disable indexes and foreign key constraints
+in your table to achieve this, which is not very practical. Still, it's a
+possibility that we should take into account.
+
+We can also use a `CREATE TEMPORARY TABLE` if our table should be dropped at the
+end of the session.
+
+### <a name="ctid"></a>But my rows don't have a unique identifier!
+
+In this case you can rename the `id` to `ctid` in our queries.
+
+`ctid` is a [system column](http://www.postgresql.org/docs/9.4/static/ddl-system-columns.html)
```

Comment:
Just replacing `id` by `ctid` in the queries above is an example, but I don't know if the postgres docs have an example like this.